python/src/spark_rapids_ml/core.py

```diff
@@ -660,6 +660,52 @@ def fitMultipleModels() -> List["_CumlModel"]:
         else:
             return super().fitMultiple(dataset, paramMaps)
 
+    def _try_stage_level_scheduling(self, rdd: RDD) -> RDD:
+        ss = _get_spark_session()
+        sc = ss.sparkContext
+
+        if ss.version < "3.4.0":
+            self.logger.warning(
+                "Stage level scheduling in spark-rapids-ml requires spark version 3.4.0+"
+            )
+            return rdd
+        elif _is_local(sc):
+            # Local mode doesn't support stage level scheduling
+            return rdd
+
+        executor_cores = sc.getConf().get("spark.executor.cores")
+        executor_gpu_amount = sc.getConf().get("spark.executor.resource.gpu.amount")
+
+        if executor_cores is None or executor_gpu_amount is None:
+            self.logger.warning(
+                "Stage level scheduling in spark-rapids-ml requires spark.executor.cores, "
+                "spark.executor.resource.gpu.amount to be set "
+            )
+            return rdd
+
+        task_gpu_amount = sc.getConf().get("spark.task.resource.gpu.amount")
+
+        if task_gpu_amount == executor_gpu_amount:
+            self.logger.warning(
+                f"The configuration of cores (exec = {executor_gpu_amount} task = {task_gpu_amount}, "
+                f"runnable tasks = 1) will result in wasted resources due to resource gpu limiting"
+                f"the number of runnable tasks per executor to: 1. Please adjust your configuration."
+            )
+            return rdd
+
+        from pyspark.resource.profile import ResourceProfileBuilder
+        from pyspark.resource.requests import TaskResourceRequests
+
+        # each training task requires cpu cores > total executor cores/2 which can
+        # ensure each training task be sent to different executor.
+        task_cores = (int(executor_cores) // 2) + 1
```

Comment:
Does this limit the number of tasks per executor to 1? 
Not quite understand the comment in line 700. Seems it indicates one executor one task.

Reply:
If an application (e.g. PCA) specifies "spark.executor.resource.gpu.amount=4", would line 701 restrict the number of GPU in use to be 1 per executor, while the application expects 4 per executor?

Reply:
Thanks for setting task_gpus to 1. Seems also need to set task_cores at line 701 to 1, so that executor.gpu.amount=4 tasks can be launched in the same executor. The current "task_cores = (int(executor_cores) // 2) + 1" restricts the #task to be 1 per executor.

Reply:
yes. good catch. thx

Reply:
Oops, No, if we change task_cores=1, task_gpus=1.0, the stage-level scheduling seems won't work. I just reverted the change.

Reply:
Finally, I got it, task_gpus means how many gpu slots the task requires, but it doesn't mean how many GPU fractions it will use. So we must use task_cpu to limit the concurrent tasks.

Reply:
@wbo4958 What is the difference between a gpu slot and GPU fraction?   What is the meaning of gpu slot, in this context?

Reply:
GPU slots are calculated according to the ResourceProfile0 (1 / task.gpu.amount), So if the task.gpu.amount is set to fraction, there'll be more than 1 GPU slot possible. While in stage-level scheduling,  "task.gpu.amount" task requires is how many gpu slots the task needs, if the value is 0~0.5, or 1, it means the task requires only 1 GPU slot,

Reply:
So if task_gpus amount in stage level scheduling is set to 1/task.gpu.amount would that also be a way to limit to 1 task per gpu for that stage?

Reply:
Not exactly, the task_gpus amount in stage-level scheduling is in 0~0.5 or 1.0, which means the task requires only 1 GPU slot. Let's assume the spark.task.resource.gpu.amount = 0.2, so there will be 5 GPU slots, In that case, even task_gpu account in stage-level is 1.0, it just requires only 1 slot instead of the whole GPU (5 slots), so the left 4 slots still can be offered to other tasks. 

In that case, we can't use task_gpu amount to limit the tasks.